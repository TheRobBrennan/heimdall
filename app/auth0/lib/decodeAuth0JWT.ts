import jwt from "jsonwebtoken"
import jwksClient from "jwks-rsa"

export const decodeAuth0JWT = async (token) => {
  return new Promise((resolve, reject) => {
    try {
      // We need to use our public keys provided by Auth0
      const jwksUri = process.env.AUTH0_DOMAIN_KEYS
      var client = jwksClient({
        jwksUri,
      })

      // Get our signing key from Auth0
      const getKey = (header, callback) => {
        client.getSigningKey(header.kid, function (err, key: Auth0SigningKey) {
          if (err) {
            console.error(`ERROR obtaining signing key from Auth0: ${err}`)
            callback(err)
          }
          var signingKey = key.publicKey || key.rsaPublicKey
          callback(null, signingKey)
        })
      }

      // Attempt to verify our Auth0 JWT
      jwt.verify(
        token,
        getKey,
        {
          algorithms: ["HS256", "RS256"],
        },
        (err, decoded) => {
          if (err) {
            console.error(`ERROR verifying Auth0 JWT: ${err}`)
            return reject(err)
          }
          return resolve(decoded)
        }
      )
    } catch (e) {
      console.error(`Unable to decode Auth0 JWT: ${token}\n\t${e}`)
      return reject(e)
    }
  })
}
