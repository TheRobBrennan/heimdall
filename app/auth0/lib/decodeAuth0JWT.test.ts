import { cleanup } from "@testing-library/react"

import { VALID_JWT_TOKEN, VALID_BUT_EXPIRED_JWT_TOKEN } from "./index"
import { decodeAuth0JWT as func } from "./decodeAuth0JWT"

describe("Our Auth0 decodeAuth0JWT lib", () => {
  afterEach(cleanup)
  describe("when supplied with a valid JWT token", () => {
    it("should receive a decoded JWT object", async () => {
      const result = await func(VALID_JWT_TOKEN)
      expect(result).toMatchSnapshot()
    })
  })
  describe("when supplied with a well-formed JWT token that has expired or cannot be verified", () => {
    it("should result in an error during processing", async (done) => {
      const result = await func(VALID_BUT_EXPIRED_JWT_TOKEN).catch((e) => {
        if (e.message === "jwt expired") {
          done()
        }
      })
    })
  })
  describe("when supplied with an unexpected string", () => {
    it("should result in an error during processing", async (done) => {
      const mockPoorlyFormedAndInvalidJWTToken = `Here is an example string that is neither a JWT token nor a valid string in the HTTP Authorization header`
      const result = await func(mockPoorlyFormedAndInvalidJWTToken).catch(
        (e) => {
          if (e.message === "jwt malformed") {
            done()
          }
        }
      )
    })
  })
})
