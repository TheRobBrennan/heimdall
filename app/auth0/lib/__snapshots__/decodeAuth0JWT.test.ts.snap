// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Our Auth0 decodeAuth0JWT lib when supplied with a valid JWT token should receive a decoded JWT object 1`] = `
Object {
  "aud": Array [
    "https://heimdall.io/demo",
    "https://heimdall-demo.us.auth0.com/userinfo",
  ],
  "azp": "6SnBQL5Zd7N1o0zysyffvaUEF5vhYJEH",
  "exp": 1612328752,
  "https://heimdall.io/jwt/claims": Object {
    "x-heimdall-allowed-roles": Array [
      "user",
    ],
    "x-heimdall-default-role": "user",
    "x-heimdall-user-id": "google-oauth2|116058668302290861810",
  },
  "iat": 1612242352,
  "iss": "https://heimdall-demo.us.auth0.com/",
  "scope": "openid profile",
  "sub": "google-oauth2|116058668302290861810",
}
`;
